pv.1
.\" $Id: pv.1,v 1.4 2008/06/07 04:54:07 kichiki Exp $
.TH PV "1" "November 4, 2007" "Version 0.9" "WaoN Manual"
.SH NAME
WaoN\-pv \- yet\-another phase vocoder
.SH SYNOPSIS
.B pv
[\fIoption \fR...]
.SH DESCRIPTION
WaoN\-pv is yet\-another phase vocoder with which you can time\-strech
and pitch\-shift a sound file.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
print this help.
.TP
\fB\-v\fR, \fB\-\-version\fR
print version information.
.PP
OPTIONS FOR FILES
.TP
\fB\-i\fR, \fB\-\-input\fR
input file (default: stdin)
.TP
\fB\-o\fR, \fB\-\-output\fR
output file in flac (default: play audio by ao)
.PP
FFT OPTIONS
.TP
\fB\-n\fR
FFT data number (default: 2048)
.TP
\fB\-w\fR \fB\-\-window\fR
0 no window
.RS
1 parzen window
.RS 0
2 welch window
.RS 0
3 hanning window (default)
.RS 0
4 hamming window
.RS 0
5 blackman window
.RS 0
6 steeper 30\-dB/octave rolloff window
.RE 1
.PP
PHASE\-VOCODER OPTIONS
.TP
\fB\-hop\fR
hop number (default: 256)
.TP
\fB\-rate\fR
synthesize rate; larger is faster (default: 1.0)
.TP
\fB\-pitch\fR
pitch shift. +1/\-1 is half\-note up/down (default: 0)
.TP
\fB\-scheme\fR
give the number for PV scheme
.RS
0 : interactive PV with curses (default)
.RS 0
1 : conventional PV
.RS 0
2 : PV by complex arithmetics with fixed hops
.RS 0
3 : Puckette's loose\-locking PV
.RS 0
4 : Puckette's loose\-locking PV by complex with fixed hops
.RS 0
5 : PV with fixed hops by Ellis
.RS 0
6 : PV in freq. domain
.RS 0
7 : plain superimpose (no-FFT)
.RE 1
.PP
.SH KEY BINDINGS IN CURSES MODE (with -scheme 0, the default)
.TP
\fBSPACE\fR        : play / stop
.TP
\fB< >\fR          : set loop range
.TP
\fB[{ }]\fR        : expand the loop range
.TP
\fBL\fR, \fBl\fR         : phase\-lock on / off
.TP
\fBW\fR, \fBw\fR         : change window
.TP
\fBH\fR, \fBh\fR         : hop\-size up / down
.TP
\fBUP\fR / \fBDOWN\fR    : pitch up / down
.TP
\fBLEFT\fR / \fBRIGHT\fR : pitch up / down
.TP
\fBR\fR, \fBr\fR, \fBHOME\fR   : reset parameters
.TP
\fBN\fR, \fBn\fR         : toggle for no-FFT mode
.TP
\fBQ\fR, \fBq\fR         : quit
.PP
.SH COPYRIGHT
Copyright \(co 1998-2008 Kengo Ichiki <kichiki@users.sourceforge.net>
Web: http://waon.sourceforge.net/
.SH "SEE ALSO"
.BR waon (1), gwaon (1)


ChangeLog
2011-12-27  Kengo Ichiki  <kichiki@users.sourceforge.net>

	* Release 0.10.

	* gwaon-about.c: 1.4 2011/12/27 14:44:46 kichiki Exp
	update.

	* gwaon-wav.c: 1.20 2011/12/27 14:39:22 kichiki Exp
	(create_wav): check ao device, and initialize,
	otherwise we got seg-fault.

	* VERSION.h:
	increment to 0.10.
	
	* main.c: 1.12 2011/12/27 13:11:00 kichiki Exp
	(main): add cleaning steps for the data.

	* midi.c: 1.7 2011/12/27 13:07:40 kichiki Exp
	(WAON_notes_output_midi): correct checks for smf_note_on/off ().

	* notes.h: 1.2 2011/12/27 13:00:56 kichiki Exp
	* notes.c: 1.2 2011/12/27 13:00:20 kichiki Exp
	add basic methods of insert, remove_at.
	add clean-up methods like regulate, remove_shortnotes,
	remove_longnotes, remove_smallnotes, remove_octaves.
	add dump methods for checking.

2008-11-30  Kengo Ichiki  <kichiki@users.sourceforge.net>

	* gwaon-about.c: 1.3 2008/12/01 00:48:29 kichiki Exp
	update.

	* gwaon-menu.c: 1.3 2008/12/01 00:38:27 kichiki Exp
	* gwaon-wav.c: 1.19 2008/12/01 00:47:59 kichiki Exp
	minor style change.

2008-11-09  Kengo Ichiki  <ichiki@tpr60>

	* gwaon-wav.c: 1.18 2008/11/09 20:59:32 kichiki Exp
	-- the following is done on 2008-06-19 --
	(draw_keyboard_hor): added.
	(draw_spectrum_frame): add horizontal keyboard.

2008-06-06  Kengo Ichiki  <ichiki@tpr60>

	* pv.c: 1.19 2008/06/07 04:54:45 kichiki Exp
	* pv.1: 1.4 2008/06/07 04:54:07 kichiki Exp
	* main.c: 1.11 2008/06/07 04:53:16 kichiki Exp
	* waon.1: 1.4 2008/06/07 04:52:47 kichiki Exp
	some corrections on manual and help documentation.

2007-11-05  Kengo Ichiki  <ichiki@tpr60>

	* waon.spec: 1.1
	initial revision.

2007-11-04  Kengo Ichiki  <ichiki@tpr60>

	* Release 0.9.

	* pv.c: 1.18 2007/11/05 02:37:10 kichiki Exp
	correct the default scheme to 0, the curses mode.

	* Makefile: 1.5 2007/11/05 02:31:44 kichiki Exp
	comment out the experimental jack stuff.

	* gwaon.1: 1.2 2007/11/05 02:29:54 kichiki Exp
	* pv.1: 1.3 2007/11/05 02:28:28 kichiki Exp
	* waon.1: 1.3 2007/11/05 02:25:22 kichiki Exp
	update the version to 0.9.

	* pv.c: 1.17 2007/11/05 02:27:00 kichiki Exp
	comment out the experimental jack stuff.
	re-order -scheme options.

	* pv-complex-curses.c: 1.7 2007/11/05 02:23:46 kichiki Exp
	widen the comment to clear the message before.

	* Makefile.waon: 1.7 2007/11/04 23:59:41 kichiki Exp
	* Makefile: 1.4 2007/11/04 23:58:16 kichiki Exp
	add notes.c for waon.

	* pv-complex.h: 1.11 2007/11/04 23:57:25 kichiki Exp
	* pv-complex.c: 1.15 2007/11/04 23:57:06 kichiki Exp
	correct some comments.
	(pv_complex_play_step): update the handling on static work area.

	* main.c: 1.10 2007/11/04 23:51:36 kichiki Exp
	use new struct WAON_notes in notes.c.

	* notes.h: 1.1 2007/11/04 23:49:55 kichiki Exp
	* notes.c: 1.1 2007/11/04 23:49:18 kichiki Exp
	added.

	* midi.h: 1.5 2007/11/04 23:48:26 kichiki Exp
	* midi.c: 1.6 2007/11/04 23:48:04 kichiki Exp
	(output_midi): removed. use WAON_notes_output_midi() instead.
	(WAON_notes_output_midi): added using struct WAON_notes in notes.c.

	* analyse.h: 1.5 2007/11/04 23:46:03 kichiki Exp
	* analyse.c: 1.8 2007/11/04 23:45:39 kichiki Exp
	(chk_note_on_off): removed. now WAON_notes_check() in notes.c is used.
	global peak_threshold is removoed.
	(init_ia_note, append_ia_note, incr_ia_note): removed.

	* fft.c: 1.9 2007/11/04 23:44:44 kichiki Exp
	(power_subtract_ave, power_subtract_octave): make the work area static.

2007-10-28  Kengo Ichiki  <ichiki@tpr60>

	* Makefile.pv: 1.7 2007/10/29 02:51:19 kichiki Exp
	* Makefile: 1.3 2007/10/29 02:50:48 kichiki Exp
	add pv-nofft.c and jack-pv.c.

	* jack-pv.h: 1.1 2007/10/29 02:49:09 kichiki Exp
	* jack-pv.c: 1.1 2007/10/29 02:48:41 kichiki Exp
	added.

	* pv.c: 1.16 2007/10/29 02:46:07 kichiki Exp
	add experimental scheme (jack and nofft).

	* pv-complex-curses.c: 1.6 2007/10/29 02:43:06 kichiki Exp
	add nofft option (just for fun).

	* pv-nofft.h: 1.1 2007/10/29 01:17:40 kichiki Exp
	* pv-nofft.c: 1.1 2007/10/29 01:17:11 kichiki Exp
	added.

	* VERSION.h: 1.7
	increment the version (for the next release).

	* gwaon-about.c: 1.2 2007/10/28 17:24:20 kichiki Exp
	update using "GTK_TYPE_ABOUT_DIALOG".

2007-10-21  Kengo Ichiki  <ichiki@tpr60>

	* pv-complex-curses.c: 1.5 2007/10/22 04:46:54 kichiki Exp
	(play_100msec_curses): update for the change in pv-complex.c.

	* gwaon-play.c: 1.8 2007/10/22 04:46:29 kichiki Exp
	(play_100msec): update for the change in pv-complex.c.

	* pv-complex.h: 1.10 2007/10/22 04:44:20 kichiki Exp
	* pv-complex.c: 1.14 2007/10/22 04:43:54 kichiki Exp
	house-keeping the code.
	(pv_complex_play_resample): change the returned value from hop_syn 
	to hop_res, the output frames.

	* hc.c: 1.7 2007/10/22 04:26:38 kichiki Exp
	(HC_complex_phase_vocoder): make the work area (tmp1, tmp2) static.

2007-10-20  Kengo Ichiki  <ichiki@tpr60>

	* gwaon.1: 1.1 2007/10/21 04:46:54 kichiki Exp
	added.

	* pv.1: 1.2 2007/10/21 04:46:20 kichiki Exp
	add informations about the key-bindings.
	add gwaon at see-also section.

	* waon.1: 1.2 2007/10/21 04:45:40 kichiki Exp
	add gwaon at see-also section.

	* gwaon-wav.c: 1.17 2007/10/21 04:44:52 kichiki Exp
	add buttons for the same functions with key-bindings.

	* pv.c: 1.15 2007/10/21 04:42:42 kichiki Exp
	(print_pv_usage): updated.

	* pv-complex-curses.h: 1.4 2007/10/21 04:04:47 kichiki Exp
	* pv-complex-curses.c: 1.4 2007/10/21 04:04:24 kichiki Exp
	(play_100msec_curses): correct the name as it is.
	(curses_print_pitch): added.

	* snd.c: 1.4 2007/10/21 04:03:30 kichiki Exp
	(sndfile_read): make the work area (buf) static.

	* gwaon-play.h: 1.2 2007/10/21 04:02:23 kichiki Exp
	* gwaon-play.c: 1.7 2007/10/21 04:01:56 kichiki Exp
	(play_100msec): correct the name as it is.

	* INSTALL: 1.3 2007/10/21 00:16:26 kichiki Exp
	updated.

	* gwaon-wav.c: 1.16 2007/10/20 20:25:37 kichiki Exp
	(create_wav): correct the initial pictch 0.

	* Makefile.pv: 1.6 2007/10/20 20:27:31 kichiki Exp
	* Makefile: 1.2 2007/10/20 20:16:42 kichiki Exp

pv.c
/* PV - phase vocoder : main
 * Copyright (C) 2007-2013 Kengo Ichiki <kengoichiki@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "memory-check.h" // CHECK_MALLOC() macro


#include "pv-complex.h"
#include "pv-conventional.h"
#include "pv-ellis.h"
#include "pv-freq.h"
#include "pv-loose-lock.h"
#include "pv-complex-curses.h"

// experimental
//#include "jack-pv.h"
#include "pv-nofft.h"

#include "VERSION.h"


void print_pv_version (void)
{
  fprintf (stderr, "WaoN-pv - yet-another phase vocoder, Version %s\n\n",
	   WAON_VERSION);
  fprintf (stdout, "Copyright (C) 1998-2013 Kengo Ichiki "
	   "<kengoichiki@gmail.com>\n");
  fprintf (stdout, "Web: http://kichiki.github.com/WaoN//\n\n");
}


void print_pv_usage (char * argv0)
{
  print_pv_version ();
  fprintf (stdout, "WaoN-pv is yet-another phase vocoder, "
	   "with which you can do time-strech \n"
	   "and pitch-shift a sound file.\n\n");
  fprintf (stdout, "Usage: %s [option ...]\n\n", argv0);
  fprintf (stdout, "Options:\n");
  fprintf (stdout, "  -h, --help\tprint this help.\n");
  fprintf (stdout, "  -v, --version\tprint version information.\n");
  fprintf (stdout, "OPTIONS FOR FILES\n");
  fprintf (stdout, "  -i, --input\tinput file (default: stdin)\n");
  fprintf (stdout, "  -o, --output\toutput file in flac"
	   " (default: play audio by ao)\n");
  fprintf (stdout, "FFT OPTIONS\n");
  fprintf (stdout, "  -n         \tFFT data number (default: 2048)\n");
  fprintf (stdout, "  -w --window\t0 no window\n");
  fprintf (stdout, "\t\t1 parzen window\n");
  fprintf (stdout, "\t\t2 welch window\n");
  fprintf (stdout, "\t\t3 hanning window (default)\n");
  fprintf (stdout, "\t\t4 hamming window\n");
  fprintf (stdout, "\t\t5 blackman window\n");
  fprintf (stdout, "\t\t6 steeper 30-dB/octave rolloff window\n");
  fprintf (stdout, "PHASE-VOCODER OPTIONS\n");
  fprintf (stdout, "  -hop       \thop number (default: 512)\n");
  fprintf (stdout, "  -rate      \tsynthesize rate; larger is faster"
	   " (default: 1.0)\n");
  fprintf (stdout, "  -pitch\tpitch shift. +1/-1 is half-note up/down"
	   " (default: 0)\n");
  fprintf (stdout, "  -scheme    \tgive the number for PV scheme\n");
  fprintf (stdout, "\t\t1 : conventional PV\n");
  fprintf (stdout, "\t\t2 : PV by complex arithmetics with fixed hops\n");
  fprintf (stdout, "\t\t3 : Puckette's loose-locking PV\n");
  fprintf (stdout, "\t\t4 : Puckette's loose-locking PV by complex"
	   " with fixed hops\n");
  fprintf (stdout, "\t\t5 : PV with fixed hops by Ellis\n");
  fprintf (stdout, "\t\t6 : PV in freq. domain\n");
  fprintf (stdout, "\t\t7 : plain superimpose (no-FFT)\n");
  fprintf (stdout, "\t\t0 : interactive PV with curses (default)\n");
  fprintf (stdout, "KEY BINDINGS IN CURSES MODE (with -scheme 0, the default)\n"
	   "\tSPACE        : play / stop\n"
	   "\t< >          : set loop range\n"
	   "\t[{ }]        : expand the loop range\n"
	   "\tL l          : phase-lock on / off\n"
	   "\tW w          : change window\n"
	   "\tH h          : hop-size up / down\n"
	   "\tUP / DOWN    : pitch up / down\n"
	   "\tLEFT / RIGHT : pitch up / down\n"
	   "\tR r HOME     : reset parameters\n"
	   "\tN n          : toggle for no-FFT mode\n"
	   "\tQ q          : quit\n");
}



int main (int argc, char** argv)
{
  char *file_in = NULL;
  char *file_out = NULL;

  /* option analysis */
  long len = 2048;
  long hop = 512;
  double rate = 1.0;
  double pitch_shift = 0.0;
  int scheme = 0;
  int flag_window = 3; // hanning window

  int i;
  for (i = 1; i < argc; i++)
    {
      if ((strcmp (argv[i], "--input" ) == 0)
	 || (strcmp (argv[i], "-i" ) == 0))
	{
	  if (i+1 < argc)
	    {
	      file_in = (char *)malloc (sizeof (char)
					* (strlen (argv [++i]) + 1));
	      CHECK_MALLOC (file_in, "main");
	      strcpy (file_in, argv[i]);
	    }
	}
      else if ((strcmp (argv[i], "--output" ) == 0)
	 || (strcmp (argv[i], "-o" ) == 0))
	{
	  if (i+1 < argc)
	    {
	      file_out = (char *)malloc (sizeof (char)
					 * (strlen (argv [++i]) + 1));
	      CHECK_MALLOC (file_out, "main");
	      strcpy (file_out, argv[i]);
	    }
	}
      else if (strcmp (argv[i], "-n" ) == 0)
	{
	  if (i+1 < argc)
	    {
	      len = (long)atoi (argv [++i]);
	    }
	}
      else if (strcmp (argv[i], "-hop" ) == 0)
	{
	  if (i+1 < argc)
	    {
	      hop = (long)atoi (argv [++i]);
	    }
	}
      else if (strcmp (argv[i], "-rate" ) == 0)
	{
	  if (i+1 < argc)
	    {
	      rate = atof (argv [++i]);
	    }
	}
      else if (strcmp (argv[i], "-pitch" ) == 0)
	{
	  if (i+1 < argc)
	    {
	      pitch_shift = atof (argv [++i]);
	    }
	}
      else if (strcmp (argv[i], "-scheme" ) == 0)
	{
	  if (i+1 < argc)
	    {
	      scheme = atoi (argv [++i]);
	    }
	}
      else if ((strcmp (argv[i], "--window") == 0)
	       || (strcmp (argv[i], "-w") == 0))
	{
	  if ( i+1 < argc )
	    {
	      flag_window = atoi (argv[++i]);
	    }
	}
      else if ((strcmp (argv[i], "--version") == 0)
	       || (strcmp (argv[i], "-v") == 0))
	{
	  print_pv_version ();
	  exit (1);
	}
      else
	{
	  print_pv_usage (argv [0]);
	  exit (1);
	}
    }

  if (file_in == NULL)
    {
      print_pv_usage (argv [0]);
      exit (1);
    }


  switch (scheme)
    {
    case 0:
      pv_complex_curses (file_in, len, hop);
      break;

    case 1:
      pv_conventional (file_in, file_out, rate, pitch_shift,
		       len, hop, flag_window);
      break;

    case 2:
      pv_complex (file_in, file_out, rate, pitch_shift,
		  len, hop, flag_window,
		  0 /* no phase lock */
		  );
      break;

    case 3:
      pv_loose_lock (file_in, file_out, rate, pitch_shift,
		     len, hop, flag_window);
      break;

    case 4:
      pv_complex (file_in, file_out, rate, pitch_shift,
		  len, hop, flag_window,
		  1 /* loose phase lock */
		  );
      break;

    case 5:
      pv_ellis (file_in, file_out, rate, pitch_shift,
		len, hop, flag_window);
      break;

    case 6:
      if (pitch_shift != 0.0)
	{
	  fprintf (stderr, "pitch-shifting is not implemented yet\n");
	  break;
	}
      pv_freq (file_in, file_out, rate, len, hop, flag_window);
      break;

    case 7:
      pv_nofft (file_in, file_out, rate, pitch_shift,
		len, hop, flag_window);
      break;
      /*
    case 8:
      pv_complex_curses_jack (file_in, len, hop);
      break;
      */

    default:
      fprintf (stderr, "invalid scheme number\n");
      break;
    }

  free (file_in);

  return 0;
}
